### ISSUE TITLE
add a short circuitry expansion feature

### ISSUE THREAD

**Issue by 0x75** [JuliaLang/julia#11782]
Hello dear julia team,

I really like the short circuit syntax. Yet i have made an observation.
SC is nice and without hassle while writing code. Reading though it becomes complicated
and there are times, when its time critical especially where i want to 'expand' the sc one liners to
python like, longer, but easier simplicity. Just to quickly understand and then change back to one-liner.

It would be cool to have such a feature added to the interpreter. 
Just select the code and... expand,  read and ..transform back.
As well as automatically transform a whole source file to beautiful code without SC and back.

best regards
0x75

**Comment by nalimilan** [comment]
Could you provide an example of what you're describing? I don't understand.

**Comment by mbauman** [comment]
This sounds like an editor/IDE feature: select the line `a && b()` and it would convert it to a three-line `if a; b(); end` block.

Are you using Juno, @0x75?  If so, you can make this suggestion over at [discuss.junolab.org](http://discuss.junolab.org).

**Comment by one-more-minute** [comment]
Things like syntax transformations and snippets are definitely nice and they're on the roadmap. This is a good idea.

I don't think this issue is particularly actionable for base Julia, though.

**Comment by ScottPJones** [comment]
Hmm, I was planning shortly on making an issue, about converting all of the:
```julia
    condition && statement # esp. things like throw, return, etc.
or
    condition || statement
```
to either
```julia
    condition &&
        statement
or
    condition ||
        statement
```
or
```julia
    if condition
        statement
    end
or
    if !condition
        statement
    end
```

The reason for this, is after looking at all of @kshyatt's excellent work improving coverage information,
it seems the current idiom basically makes it impossible to get really good coverage info, because the coverage tools work on a line basis, not on a basic block basis.
I don't think this really should be just an editor expansion, I think we need to talk about what the best idiom for use in Julia code is, taking into account the issues of coverage accuracy, conciseness, and ease of understanding the code (esp. for newcomers).

**Comment by 0x75** [comment]
sorry for the very late reply. 
I wrote short circuits but i mean a larger field. I do not program in julia mainly so i will explain it using little helpers like c, python and so on. I saw a early developer talk of one of the (senior?) designer who discussed the julia idom. he told the audience about a discussion about how the idiom has to look and how he doesnt like the now idiomatic short circuit syntax and the c-like "cryptic" one liners. He stated he likes the pythonic readability first approach more. I like slick and efficient code but i totally see the issue of readability. If in a hurry this things really slow you down understanding foreign code. My suggestion tries to evade the (futile ) idiom discussion and solve it with a feature for the interactive julia shell (so not an editor thing). The idea is write slick code whenever you want. And whenever you want to have a slick reading experience you highlight the line + press a key and the line(s) get formatted into "python like" clear readable code. As soon as you understood the part you press the key again an the line becomes "optimized" again. This works the other way around too. Highlight your code and "auto optimize" it into slickness. Which would be a great learning feature for beginners and intermediates to  write better code. this would enforce a kind of coding style while maintaining fast accessability. While not easy to implement it would be a really helpful feature.

**Comment by StefanKarpinski** [comment]
This is not a language feature, it's an IDE feature.

**Comment by 0x75** [comment]
Well it is what you make of it. The python pep8 thing can be seen as somewhat of a hybrid if you look at it generously (just for the sake of catching my drift). And my suggestion is just one step further. A feature to a linter to "beautify" part of the code in this way. Implemented as an isolated IDE feature it is basically worthless since every IDE is implemented different. Implementing it as a julia shell builtin listing feature would set a standard.  I think it might have a bigger effect than just being an IDE feature (which is my point by the way) it might be a feature that is seen as a language feature. From the point of the user it would be seen as a language feature which would add to the lovely specialness and usefulness of the language. I am pointing out that since julia is one of the languages with a shell this could be implemented apart from an IDE. I think that more features that will make programming easier can be implemented in this way. In my point of view this is a design decision, since such features can be implemented as a part of the julia program (like the shell or the package manager. not as part of the language itself if you want to be precise).  Honestly i think it would be a waste to leave such kind of features to the IDEs. IDEs as well could just "call the feature" from the "julia package" and such access a unified interface with unique useful features.
I have made several similar suggestions of this "blending" features to other projects which turned out successful (nobody seems to have thought of direct asm editing until i suggested it to the hopper developer (the hopper disassembler), package managing/dependency checking for c/c++ and jit-compilation (not by me) are other examples). The larger picture is that the julia user base can develop similar features like that. This would take the initiative away from the IDE projects which are not very impressing and slowly evolving in my opinion. I see it as a growth and development vector to "the language". I am aiming at a dynamic analog to the text-mate plugin ecosystem where the user base has proven to be quite productive. Developers would need a defined playground and (plug-in) interface for such features. They might come up with things no one has thought of or wished for for a long time (just like multiline editing in sublime text).  
Excuse the lengthy description, writing about something new and undefined is kind of difficult.
